### ISSUE TITLE
Group by never completes

### ISSUE THREAD

**Issue by rikuiki** [duckdb/duckdb#9718]
### What happens?

I am testing new duckdb function of larger than memory aggregations.
My query finishes for 2B rows in 1m30s, but never completes for 3B rows.

Query execution progress reaches 49% while doing heavy IO operations in few minutes, and then I see that duckdb process constantly uses 20 cores, and doesn't produce any IO and progress never changes, I waited for 12h for completion.

### To Reproduce

```sql
.open t
.timer on

SET memory_limit='16GB';
SET threads TO 30;

create table t as
select
  cast(v.range % 300000000 as varchar) v
from range(3000000000) v;

create table t2 as select v, count(*) from t group by 1;
```

### OS:

Debian Linux

### DuckDB Version:

0.9.1

### DuckDB Client:

CLI

### Full Name:

Riku Iki

### Affiliation:

nextintelligent.com

### Have you tried this on the latest `main` branch?

I have not tested with any build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by l1t1** [comment]
out of memory, so it use swap file, it's much slower.

**Comment by rikuiki** [comment]
> out of memory, so it use swap file, it's much slower.

Server has much more memory than 16GB limit set.
duckdb process still uses just 16GB.
Server has swap disabled.
I don't see any IO traffic for the duckdb process and very little IO traffic in general on that server.

**Comment by l1t1** [comment]
if you unset temp_directory, will the sql raise an error?

**Comment by rikuiki** [comment]
> if you unset temp_directory, will the sql raise an error?

How can I set it? If I set it temp_directory='', it will use current dir?..

Also, one correction, my query produces heavy IO (2GB/s) up until 49% of progress in a first few mins of execution. Then it stuck in CPU with no IO traffic for indefinite time.

**Comment by michaelmdresser** [comment]
Hi @rikuiki I think I may have encountered something similar in https://github.com/duckdb/duckdb/issues/8731, specifically https://github.com/duckdb/duckdb/issues/8731#issuecomment-1712097373

> htop reports the process is using about 1 CPU core (the column with 118), with near-zero variation in the usage:

Other times I've encountered this problem at higher data scales I have seen multi-core usage that was perfectly pegged at 100% on each core in a similar fashion.

**Comment by l1t1** [comment]
`reset temp_directory` to not use temp files

**Comment by rikuiki** [comment]
> `reset temp_directory` to not use temp files

Ok, I reset that setting, and query which creates temp table (before grouping by) dies with OOM.

**Comment by l1t1** [comment]
try 
create table t as
select
cast(v.range % 300000000 as uint32) v
from range(3000000000) v;

**Comment by rikuiki** [comment]
> try create table t as select cast(v.range % 300000000 as uint32) v from range(3000000000) v;

My target dataset is strings, not int.

**Comment by lnkuiper** [comment]
Tuples need a null mask (1 byte), string (16 bytes), hash (8 bytes), `count` aggregate state (8 bytes), for a total of 33 bytes. We align this to the next 8, so that's 40 bytes per tuple. We also need ~12 bytes per tuple in the hash table, which comes down to ~52 bytes per tuple.

Given your 3B input rows, this is ~156GB. Of course, each tuple appears 10 times, so we'd actually only need 15.6GB, but DuckDB can only efficiently de-duplicate groups if they appear close to each other in the data ("interesting orderings", which most real-world data has). So, the data is only de-duplicated at the very end, which causes a lot of I/O.

So, DuckDB materializes ~156GB of data, which, with 16GB of memory, causes ~140GB of writing and reading. Depending on your disk speed, this may take a while.

That being said, I'm running this on my 2020 MacBook Pro, which has a very fast disk, but it still takes a long time. I've looked into why this happens and pushed a small fix in #9946, which should improve performance here. Now, the query completes in ~4 minutes on my laptop, which has 8 threads / 16GB of memory.

**Comment by rikuiki** [comment]
Hi team, thank you for looking at this. Fix indeed fixes the problem for 3B rows, but issue now is reproducible at 10B rows, e.g. when I generate data as:

create table t as
select
  cast(v.range % 1000000000 as varchar) v
from range(10000000000) v;

Some comments 

> So, DuckDB materializes ~156GB of data, which, with 16GB of memory, causes ~140GB of writing and reading. Depending on your disk speed, this may take a while.

I have NVME raid, and DuckDB pushes 5GB/s read/writes, so this shouldn't be a problem.

However, what I see is that DuckDB reaches 49% of progress, and then after some time doesn't produces any IO, but stuck in some CPU loop while loading 10-20 cores with 100% forever (I waited for 6h with no changes in progress).

**Comment by rikuiki** [comment]
#9989 fixed this, @szarnyasg fyi

**Comment by kevfly16** [comment]
I have a similar issue with a different query and dataset. My progress bar gets forever stuck at 34%. Manually limiting the thread count to 32 threads seems to magically solve it.

**Comment by singlespark6979** [comment]
> Tuples need a null mask (1 byte), string (16 bytes), hash (8 bytes), `count` aggregate state (8 bytes), for a total of 33 bytes. We align this to the next 8, so that's 40 bytes per tuple. We also need ~12 bytes per tuple in the hash table, which comes down to ~52 bytes per tuple.
> 
> Given your 3B input rows, this is ~156GB. Of course, each tuple appears 10 times, so we'd actually only need 15.6GB, but DuckDB can only efficiently de-duplicate groups if they appear close to each other in the data ("interesting orderings", which most real-world data has). So, the data is only de-duplicated at the very end, which causes a lot of I/O.
> 
> So, DuckDB materializes ~156GB of data, which, with 16GB of memory, causes ~140GB of writing and reading. Depending on your disk speed, this may take a while.
> 
> That being said, I'm running this on my 2020 MacBook Pro, which has a very fast disk, but it still takes a long time. I've looked into why this happens and pushed a small fix in #9946, which should improve performance here. Now, the query completes in ~4 minutes on my laptop, which has 8 threads / 16GB of memory.

Sorry to interrupt your discussion. I'm wondering how DuckDB handle out-of-core execution of group by aggregation.

Since TupleDataLayout won't take care of aggregate field, what if the aggregate function return a variable length data?

The aggregate update function will detroy and copy string with delete and new operator.

Even though group by keys can be abondoned and swap to disk, the aggregate result still in memory.

Here is my testing sql. 40M records, 10M cardinality.

SET memory_limit='100MB';
SET threads TO 1;

create table t as
select v.range % 10000000 vi,
cast(v.range % 40000000 as varchar) || 'aaaaaaaaaaaaaaaaa' v
from range(40000000) v;

select vi, max(v) from t group by 1;
